### DISCUSSION TITLE
pyMOR and other CFD solvers

### DISCUSSION THREAD

**Discussion by jacktang** [pymor/pymor#1642] [Q&A]
Hello,

I am new to pyMOR and want to build ROM using pyMOR. But my train data is dumped from outside solver, like  openFOAM and code_saturne. My question is how can I convert the dumped dataset to pyMOR training set? Does the community has some best  practise / guides related this?

**Comment by sdrave** [comment]
Hi, @jacktang! If you have your training data on disk, all you need to do is to find a way to make it available as a NumPy array. How that works will depend on the file format in which your solver has stored the data. Then you can convert the NumPy array to a pyMOR `VectorArray` by writing

```python
U = NumpyVectorArray.from_numpy(the_data)
```

where `the_data` has to be a 2D NumPy array, where the individual data vectors are stored as row vectors.

Be aware that with only the training data you are limited to purely data-driven MOR methods like DMD or the neural network based approaches.

**Comment by jacktang** [comment]
@sdrave Thanks for your response!

I am solving a time-dependent cylinder wake (2d) problem, Besides the training data(part of solution data), I also have the mesh and solution data. So I suppose the pseudocode looks like

```
def make_fom(mesh_file, solution_data)
   mesh = load_mesh(mesh_file)
   solution = load_solution(solution_data)
   fom = SomeParametricModel(...)
   return fom
```

I don't know to assemble `fom` using mesh and the solution data. Can you give me one example?

> Be aware that with only the training data you are limited to purely data-driven MOR methods like DMD or the neural network based approaches.

And thanks for the hints! I am going to read the paper about DMD. <s> Does the community has the examples or tutorials related to DMD?  </s> Well, I get the demo: dmd_identification.py

**Reply by renefritze** [reply]
> I don't know to assemble fom using mesh and the solution data. Can you give me one example?

Have a look at [creating models](https://docs.pymor.org/2021-2-0/technical_overview.html#creating-models). It refers to functions in pyMOR that setup models for the builtin analytical problems. 

Some more examples for model creation: [1](https://docs.pymor.org/2021-2-0/tutorial_external_solver.html#putting-it-all-together) [2](https://docs.pymor.org/2021-2-0/tutorial_optimization.html#an-elliptic-model-problem-with-a-linear-objective-functional)

**Reply by sdrave** [reply]
Also note that for a purely data-driven approach, you do not really need the mesh except for visualization, which should be handled by the CFD code.

**Reply by jacktang** [reply]
@renefritze @sdrave Thanks for your suggestions!
I have read `Binding an external PDE solver to pyMOR`. I think my case is different from the tutorial. The tutorial tells me how to assemble external solver with pyMOR. The solver and pyMOR are coupled together, and in my case pyMOR and CFD solver are independent. The CFD team only submit solution data to us.

And my question is that does it still need to define operator/rhs/products even I have the solution space data? How can I pass solution data to model?

**Comment by jacktang** [comment]
Well, I tried the data-driven solution in current days.
My DOE is quite simple: I adopt the existed problem presented in tutorial, and I know the exact `fom`(fom0), and collect training data from the solution space. Then pass the training data to the `NeuralNetworkReductor` where `NeuralNetworkModel` acts as `"fom"`(fom1). Finally I compare the result of `rom.solve` and `fom0.solve` . But the result seems not so good. Can you tell me what's wrong in my code ? Appreciate your time!


```
from pymor.basic import *
from pymor.core.config import config
import numpy as np

config.HAVE_TORCH = True

from pymor.reductors.neural_network import NeuralNetworkReductor
from pymor.models.neural_network import  (FullyConnectedNN, NeuralNetworkModel)

problem = StationaryProblem(
    domain=RectDomain(),

    rhs=LincombFunction(
        [ExpressionFunction('10', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.1]),

    diffusion=LincombFunction(
        [ExpressionFunction('1 - x[0]', 2), ExpressionFunction('x[0]', 2)],
        [ProjectionParameterFunctional('mu'), 1]),

    dirichlet_data=LincombFunction(
        [ExpressionFunction('2 * x[0]', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.5]),

    name='2DProblem'
)

fom, _ = discretize_stationary_cg(problem, diameter=1/50)

parameter_space = fom.parameters.space((0.1, 1))
training_set = parameter_space.sample_uniformly(1000)
validation_set = parameter_space.sample_randomly(20)

d_in = 1
d_out = fom.solution_space.dim # 5101
neural_network = FullyConnectedNN([d_in, 3 * (d_in + d_out), 3 * (d_in + d_out), d_out]).double()

nn_model = NeuralNetworkModel(neural_network, Parameters(mu=1))

reductor = NeuralNetworkReductor(nn_model,
                                 training_set,
                                 validation_set,
                                 l2_err=1e-5,
                                 ann_mse=1e-5)
rom = reductor.reduce(restarts=100)

test_set = parameter_space.sample_randomly(10)

U = fom.solution_space.empty(reserve=len(test_set))
U_nn = fom.solution_space.empty(reserve=len(test_set))
U_red = fom.solution_space.empty(reserve=len(test_set))

speedups = []

import time

for mu in test_set:
    tic = time.perf_counter()
    U.append(fom.solve(mu))
    time_fom = time.perf_counter() - tic

    tic = time.perf_counter()
    U_red.append(reductor.reconstruct(rom.solve(mu)))
    time_red = time.perf_counter() - tic

    tic = time.perf_counter()
    U_nn.append(nn_model.solve(mu))
    time_red = time.perf_counter() - tic

    speedups.append(time_fom / time_red)

absolute_errors = (U - U_red).norm()
relative_errors = (U - U_red).norm() / U.norm()

print("absolute errors:", np.average(absolute_errors))
print("relative errors:", np.average(relative_errors))
```

The output is 

```
absolute errors: 105.06361731964853
relative errors: 1.00669761493105
```

**Reply by HenKlei** [reply]
Hi @jacktang! The main problem here is actually that the `NeuralNetworkReductor` takes as first argument the FOM and not the `NeuralNetworkModel`. When calling `reduce` on the `NeuralNetworkReductor`, what you get as `rom` is a `NeuralNetworkModel`. So your code should work if you replace
```
reductor = NeuralNetworkReductor(nn_model,
                                 training_set,
                                 validation_set,
                                 l2_err=1e-5,
                                 ann_mse=1e-5)
```
by
```
reductor = NeuralNetworkReductor(fom,
                                 training_set,
                                 validation_set,
                                 l2_err=1e-5,
                                 ann_mse=1e-5)
```
Internally, the `NeuralNetworkReductor` calls `fom.solve(mu)` for every `mu` in `training_set` and `validation_set` to create training data. Therefore, you might want to reduce the number of training samples, because solving the FOM for 1000 training instances is typically too costly. Further, the input and the output dimension of the neural network are computed automatically in the `NeuralNetworkReductor`. The input dimension is the dimension of the parameter space, whereas the output dimension is the dimension of the **reduced** space, so not known beforehand (except you prescribe the reduced basis space dimension). For more details, take a look at [the documentation](https://docs.pymor.org/2021-2-0/autoapi/pymor/reductors/neural_network/index.html#pymor.reductors.neural_network.NeuralNetworkReductor), in particular the `reduce` function of the `NeuralNetworkReductor`.

**Comment by HenKlei** [comment]
Maybe to also answer your original question about other CFD solver: Thus far, it is necessary to create a model that can serve as `fom` and be passed to the `NeuralNetworkReductor`. Basically, this model only has to implement how `solve` is performed for a parameter `mu`. However, if you only have snapshot data and no Python code to run the simulations for new parameters (I assume that this is the case you are asking for), then it is currently not possible to directly use the `NeuralNetworkReductor`. I already discussed this issue with @sdrave and he proposed to change the `NeuralNetworkReductor` accordingly, such that it also accepts snapshots with the corresponding parameters as input. There are some details to address, but I will try to come up with a solution as soon as possible.

**Reply by jacktang** [reply]
@HenKlei Thanks for the informative reply!

> Basically, this model only has to implement how solve is performed for a parameter mu. 

Yes, I am going to create some model called `SolutionReadyModel`which takes `solution` as input parameter, and it implements `_compute_solution` method.  PyMOR is still new to me, so the implementation will take some time.

> However, if you only have snapshot data and no Python code to run the simulations for new parameters (I assume that this is the case you are asking for)

Yes, it is the case I am facing now.

**Reply by HenKlei** [reply]
> @HenKlei Thanks for the informative reply!

You are welcome!

> > Basically, this model only has to implement how solve is performed for a parameter mu.
> 
> Yes, I am going to create some model called `SolutionReadyModel`which takes `solution` as input parameter, and it implements `_compute_solution` method. PyMOR is still new to me, so the implementation will take some time.

Just to clarify this: `_compute_solution` takes a **parameter** `mu` as input and is supposed to **compute** the corresponding solution for that parameter. If you only have snapshots, the `_compute_solution` method could only be evaluated for those parameters mu for which the corresponding snapshot is available. It would be possible to implement it this way, since you can pass the parameters for which the solution is known as `training_set` and `validation_set` to the `NeuralNetworkReductor`. However, this does not feel right to me, since you really have to make sure that your model is only evaluated for those parameters for which the snapshots are available.

**Reply by jacktang** [reply]
Hello @HenKlei !

> However, this does not feel right to me, since you really have to make sure that your model is only evaluated for those parameters for which the snapshots are available.
 
These days I read more codes about pyMOR. I think in pure data driven solution, `training_set` and `validation_set` should be split from snapshots. Here is my definition of `SolutionReadyModel`

```
from pymor.models.interface import Model

class SolutionReadyModel(Model):
    def __init__(self, snapshots, paramters={}, output_functional=None,
                    products=None, error_estimator=None, visualizer=None, name=None):

        super().__init__(products=products, error_estimator=error_estimator,
                            visualizer=visualizer, name=name)

        self.__auto_init(locals())
        self.parameters = paramters
        self.solution_space = snapshots
        self.output_functional = output_functional or ZeroOperator(NumpyVectorSpace(0), self.solution_space)
        assert self.output_functional.source == self.solution_space
        self.dim_output = self.output_functional.range.dim

    def _compute_solution(self, mu=None, **kwargs):

        idx = np.where(self.parameters == mu)
        U = self.solution_space[idx]

        return U
```

And the usage

```
training_set = parameter_space.sample_uniformly(100)
validation_set = parameter_space.sample_randomly(2)

mu_col = training_set.copy()
mu_col.append(validation_set)

snapshots = fom.solution_space.empty(reserve=(len(training_set) + len(validation_set)))

for mu in training_set:
    snapshots.append(fom.solve(mu))

for mu in validation_set:
    snapshots.append(fom.solve(mu)) 

fom = SolutionReadyModel(snapshots, mu_col)
reductor = NeuralNetworkReductor(fom,
                                 training_set,
                                 validation_set,
                                 l2_err=1e-5,
                                 ann_mse=1e-5)
```

But it throws error

```
----> [9] fom = SolutionReadyModel(snapshots, mu_col)
     [10] reductor = NeuralNetworkReductor(fom,
     [11]                                 training_set,
     [12]                                 validation_set,
     [13]                                 l2_err=1e-5,
     [14]                                 ann_mse=1e-5)
     [15] rom = reductor.reduce(restarts=100)

File ~/opt/miniconda3/lib/python3.9/site-packages/pymor/core/base.py:280, in ImmutableMeta._call(self, *args, **kwargs)
    279 def _call(self, *args, **kwargs):
--> 280     instance = super().__call__(*args, **kwargs)
    281     assert all(hasattr(instance, arg) for arg in instance._init_arguments), \
    282         (f'__init__ arguments {[arg for arg in instance._init_arguments if not hasattr(instance,arg)]} '
    283          f'of class {self.__name__} not available as instance attributes\n'
    284          f'(all __init__ args need to be attributes for with_ to work).')
    285     instance._locked = True
...
---> 49 dict.__init__(new, sorted(dict(*args, **kwargs).items()))
     50 new._post_init()
     51 return new

ValueError: dictionary update sequence element #0 has length 1; 2 is required
```

How can I fix this? Thanks!

**Reply by HenKlei** [reply]
During the last developer meeting we decided that it should also be possible to use the `NeuralNetworkReductor` in a purely data-driven manner. Therefore, I will look into this in the next days. This will certainly also be helpful in your case, such that you can directly feed your data to the reductor and no full-order model is required. I will provide some sample code here for you as soon as the changes made their way to `main`.
